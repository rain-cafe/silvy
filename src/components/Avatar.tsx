import styles from './Avatar.module.scss';
import Image from 'next/image';

type AvatarProps = {
    src: string;
}

export default function Avatar({
    src
}: AvatarProps) {
    return (
        <Image
            className={styles.avatar}
            width={40}
            height={40}
            src={src}
        />
    )
}