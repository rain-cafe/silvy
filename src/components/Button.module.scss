.button {
    display: inline-flex;
    align-items: center;
    border-radius: 5px;
    border: none;
    cursor: pointer;
    color: inherit;

    background: transparent;
    transition: 0.15s ease-in-out;
    transition-property: background-color, color;

    &:active {
        color: inherit;
    }

    &.sz-small {
        padding: 10px;
        font-size: 16px;
    }

    &.sz-medium {
        height: 50px;
        padding: 0 10px;
        font-size: 24px;
    }

    &.sz-large {
        height: 50px;
        padding: 0 20px;
        font-size: 30px;
        font-weight: bold;
    }

    &:hover {
        &.an-darken {
            background-color: rgba(0, 0, 0, 0.25);
        }

        &.an-swap {
            // Figure out if there's a way of clipping the text to be transparent
            background-color: #f492a5;
            color: #161b21;
        }
    }
}