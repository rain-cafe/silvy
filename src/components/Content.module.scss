.content {
    display: flex;
    gap: 10px;

    &.main {
        margin: auto;
        max-width: 1480px;
        height: 100%;
        width: 100%;
        background-color: rgba(0, 0, 0, 0.25);
        border: 1px solid rgba(0, 0, 0, 0.25);
        border-top: none;
        border-bottom: none;
    }

    &.dir-horizontal {
        flex-direction: row;
        align-items: center;
        justify-content: center;
    }

    &.dir-vertical {
        flex-direction: column;
    }
}