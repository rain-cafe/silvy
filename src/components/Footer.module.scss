.footer {
    height: 60px;
    padding-bottom: 20px;

    img {
        margin-left: 0.5rem;
    }
    
    a {
        display: flex;
        justify-content: center;
        align-items: center;
    }
}

.poweredBy {
    display: flex;
    align-items: center;
    gap: 10px;
    color: white;
}

.logo {
    height: 1em;
}