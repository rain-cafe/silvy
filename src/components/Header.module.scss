.header {
    display: flex;
    align-items: center;
    justify-content: flex-start;
    height: 70px;
    box-sizing: border-box;
    background: #161b21;
}

.content {
    padding-left: 20px;
    padding-right: 20px;
}

.name {
    font-family: 'Dancing Script', cursive;
}

.discord {
    font-family: 'Whitney', 'Helvetica Neue', Helvetica, Arial, 'Lucida Grande', sans-serif;
}