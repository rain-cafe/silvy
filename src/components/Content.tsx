import { Direction } from '../constants/components';
import { classnames } from '../utils/classnames';
import styles from './Content.module.scss';

interface ContentProps {
    children: React.ReactNode;
    className?: string;
    direction?: Direction
    main?: boolean;
}

export function Content({
    children,
    className,
    direction = Direction.VERTICAL,
    main = false
}: ContentProps) {
    return (
        <div className={classnames(
            styles.content,
            styles[`dir-${direction}`],
            main && styles.main,
            className
        )}>
            {children}
        </div>
    );
}