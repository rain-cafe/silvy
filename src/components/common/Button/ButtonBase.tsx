import type { PolymorphicPropsWithoutRef } from 'react-polymorphic-types';
import classNames from 'classnames';
import Link from 'next/link';
import { useReadOnlyCachedState } from '../../hooks/use-cached-state';
import styles from './ButtonBase.module.scss';
import React from 'react';
import { Sizes } from '../../../constants/components';

export const DEFAULT_ELEMENT = 'button';

export interface ButtonBaseOwnProps {
    children: React.ReactNode;
    className?: string;
    disabled?: boolean;
    size?: Sizes;
}

export type ButtonBaseProps<
    T extends React.ElementType = typeof DEFAULT_ELEMENT
> = PolymorphicPropsWithoutRef<ButtonBaseOwnProps, T>;

export function ButtonBase<E extends React.ElementType = typeof DEFAULT_ELEMENT>({
    as,
    className: externalClassName,
    children,
    disabled = false,
    size = Sizes.SMALL,
    ...extraProps
}: ButtonBaseProps<E>) {
    // eslint-disable-next-line @typescript-eslint/naming-convention
    const Element: React.ElementType = as || DEFAULT_ELEMENT;
    const className = useReadOnlyCachedState(() => {
        return classNames(
            styles.buttonBase,
            styles[`sz-${size}`],
            externalClassName,
            {
                [styles.disabled]: disabled
            }
        );
    }, [disabled, externalClassName]);

    if (Element === Link) {
        return (
            <Element
                {...extraProps}
                className={className}
                disabled={disabled}
                draggable={false}
            >
                <a
                    data-testid='button-base'
                    className={className}
                >
                    {children}
                </a>
            </Element>
        );
    }

    return (
        <Element
            data-testid='button-base'
            {...extraProps}
            className={className}
            disabled={disabled}
            draggable={false}
        >
            {children}
        </Element>
    );
}