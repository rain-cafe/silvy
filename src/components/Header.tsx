
import { BsPatchPlus, BsPatchPlusFill, BsPeopleFill, BsBookFill } from 'react-icons/bs';
import styles from './Header.module.scss';
import { Button } from './Button';
import { Alignment, Direction, Sizes } from '../constants/components';
import { Content } from './Content';
import Link from 'next/link';
import Login from './Login';
import { IconButton } from './IconButton';
import { Popover, PopoverItem } from './Popover';
import { useSession } from 'next-auth/react';

export function Header() {
    const { data: session } = useSession();

    return (
        <div className={styles.header}>
            <Content className={styles.content} direction={Direction.HORIZONTAL} main>
                <Button
                    as={Link}
                    className={styles.name}
                    size={Sizes.LARGE}
                    href='/'
                >
                    Silvy
                </Button>
                <div style={{
                    flex: 1
                }}></div>
                {session && (
                    <Popover
                        toggle={<IconButton
                            icon={BsPatchPlus}
                            hoverIcon={BsPatchPlusFill}
                            size={30}
                        />}
                    >
                        <PopoverItem>
                            <BsBookFill size={20} />
                            Create a Strat
                        </PopoverItem>
                        <PopoverItem>
                            <BsPeopleFill size={20} />
                            Create a Team
                        </PopoverItem>
                    </Popover>
                )}
                <Login />
            </Content>
        </div>
    );
}