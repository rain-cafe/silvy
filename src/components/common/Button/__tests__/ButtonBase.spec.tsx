import Link from 'next/link';
import React from 'react';
import { createRenderComponent } from '@rain/test-utils/render';
import { ButtonBase, ButtonBaseProps } from '../ButtonBase';
import { Sizes } from '../../../../constants/components';
import styles from '../ButtonBase.module.scss';

describe('Component(Button)', () => {
    const renderButton = createRenderComponent<ButtonBaseProps<React.ElementType>>(ButtonBase, {
        children: 'UwU'
    });

    it('should display the Button', () => {
        const { baseElement } = renderButton()

        expect(baseElement).toMatchSnapshot()
    });

    it('should support overriding the tag type', () => {
        const { baseElement } = renderButton<ButtonBaseProps<'a'>>({
            as: 'a'
        });

        expect(baseElement.querySelector('a')).toBeTruthy();
    });

    it('should support overriding the size', () => {
        const { getByTestId, expectedProps } = renderButton({
            size: Sizes.MEDIUM
        });

        expect(getByTestId('button-base')).toHaveClass(styles[`sz-${expectedProps.size}`]);
    });

    it('should support NextJS Links', () => {
        const { baseElement, getByTestId } = renderButton<ButtonBaseProps<typeof Link>>({
            as: Link,
            href: 'https://silvy.rains.cafe'
        });

        expect(baseElement.querySelector('a')).toBeTruthy();
        expect(getByTestId('button-base')).toHaveAttribute('href');
    });

    it('should support being enabled', async () => {
        const { getByTestId, user, expectedProps } = renderButton({
            onClick: jest.fn()
        });

        await user.click(getByTestId('button-base'));

        expect(expectedProps.onClick).toHaveBeenCalledTimes(1);
    });

    it('should support being disabled', async () => {
        const { getByTestId, user, expectedProps } = renderButton({
            disabled: true,
            onClick: jest.fn()
        });

        await user.click(getByTestId('button-base'));

        expect(expectedProps.onClick).toHaveBeenCalledTimes(0);
    });
});
