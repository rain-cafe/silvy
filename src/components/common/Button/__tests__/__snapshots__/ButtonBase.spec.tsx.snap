// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Component(Button) should display the Button 1`] = `
<body>
  <div>
    <button
      class="buttonBase sz-small"
      data-testid="button-base"
      draggable="false"
    >
      UwU
    </button>
  </div>
</body>
`;
